### ISSUE TITLE
[OPTIMIZATION] Potential fix to incorrect SOC Behaviour

### ISSUE THREAD

**Issue by chrischang5** [UBC-Solar/Simulation#60]
# Background

Initially, our optimization was assumed to have incorrect charging behaviour. For background on this, our simulation currently does an initial run with a "guess" of an input speed array and produces some distance result without performing any optimization. After, it starts to optimize/maximize the distance value calculated by the simulation. 

Early runs of the simulation had the "guess" showed charging behaviour (See SOC-time plot):
![clear_soc_charge](https://user-images.githubusercontent.com/64388914/120903056-c275e680-c5f8-11eb-9904-ef7d1f366b52.png)

However, the optimized result showed no signs of charging behaviour and dropped off rapidly (See SOC-time plot):
![show_justin_2](https://user-images.githubusercontent.com/64388914/120903065-d6214d00-c5f8-11eb-9b9c-b4c442759b8c.png)

# Findings

This may just be luck, but some results indicate that it's not any problem rooted in simulation physics models. Viewing the SOC-time plot in the photo below, there is charging behaviour between t = 4 h to t = 4.5 h. Also, note the small amounts of delta energy between the same period. This is a good sign since it proves charging behaviour is possible in the optimization stage and some other thing is causing the SOC to decrease really fast.  

![image](https://user-images.githubusercontent.com/64388914/120903118-20a2c980-c5f9-11eb-8da2-77eed3b1d19f.png)

This is my naively hopeful explanation of the above diagram:
Green Highlighted Region: Charging from 7 AM-9 AM
Orange Region: Decrease in SOC due to low amounts of sun and travel at moderate speed 
Light Blue Region: Travel at a LOW speed and nearing the noon peak of solar irradiance. This leads to charging behaviour.
Red Arrow Region: The car simply goes too fast despite the most amount of sun present. This causes the battery to drain out completely.

Put simply, my hypothesis is fast = lose energy and slow = gain energy. Not sure if that's a breakthrough discovery in itself, but it certainly eliminates the possibility of a broken physics model. Analyzing the Delta Energy-time plot for the first two photos also aligns up with this type of thinking. A huge speed means a huge negative spike in delta energy and slower speeds mean smaller ripples in the Delta Energy curve.

# Potential Solutions
It's actually hard to say if this problem actually needs fixing... But if we're looking to get a better result on our models. we could:

- Lower the guess' upper bound.
- See #51 for improving the battery model
- Analyze the efficiency values given by our motor/motor controller efficiency values
- Start simulating with a higher initial SOC

**Comment by chrischang5** [comment]
Some comments I had while writing this:

Because high speeds drain the battery very quickly, this raises some questions about how fast the car can sustainably go. Can adding a different discharge curve as per #51 fix it? It will all depend on if the discharging rate is slower than the charging rate.

I'm also interested to see how much of an effect changing the initial battery charge from 0.6 to 1 will have. I predict it won't run out by the end of the 10 hours, but the speed array could look different in this case and I could be totally wrong.

**Comment by chrischang5** [comment]
Another consideration is that do we care if the battery dies at the end of a ~10 h simulation? We just have to ensure we are able to make it to the end of the driving period and then we can charge.

I've heard Alex talk about strategies where teams start with 1 then end at 0.4, charge to 0.8, and end at 0.2, then go to the last race day with 0.6 and end with 0.

Maybe our optimization will automatically generate that result. If it is the case, that would be really cool to see, but I'm not sure how likely that is.

**Comment by chrischang5** [comment]
Resolved. I'm pretty certain my explanation is correct.
